cpp/FileTransferAgent.cpp

```diff
@@ -398,48 +403,66 @@ RemoteStorageRequestOutcome Snowflake::Client::FileTransferAgent::uploadSingleFi
   m_FileMetadataInitializer.initEncryptionMetadata(fileMetadata);
   CXX_LOG_TRACE("Encryption metadata init done");
 
-  std::basic_iostream<char> *srcFileStream;
-  ::std::fstream fs;
-
-  if (m_uploadStream)
+  RemoteStorageRequestOutcome outcome = RemoteStorageRequestOutcome::SUCCESS;
+  do
   {
-    srcFileStream = m_uploadStream;
-  }
-  else
- {
-    try {
-      fs = ::std::fstream(fileMetadata->srcFileToUpload.c_str(),
-                          ::std::ios_base::in |
-                          ::std::ios_base::binary);
-    }
-    catch(...)
+    if (retry > 0)
     {
-      std::string err= "Could not open source file " + fileMetadata->srcFileToUpload ;
-      throw SnowflakeTransferException(TransferError::FAILED_TO_TRANSFER, err.c_str());
+      unsigned long waitInMicroSecs = (retry << 1) * 100000 ; //First retry 200ms, Second retry 400ms, Third retry 800ms
+      CXX_LOG_DEBUG("Retry count %d, Retrying put file %s.", retry, fileMetadata->srcFileName.c_str());
```

Comment:
Before the sleep, send a message to telemetry stating that the sleep, anf the current outcome.

Reply:
put failure msg's are sent to ODBC layer and sent to inband telemetry. libsnowflakeclient does not have inband telemetry and cannot directly send inband events.